src/main/java/app/linguistai/bmvp/service/gamification/StoreService.java

```diff
@@ -0,0 +1,206 @@
+
+            // Save the user item
+            userItemRepository.save(userItem);
+
+            log.info("User {} purchased item {} successfully", email, storeItem.getType());
+            return RUserItem.builder().userItem(userItem).build();
+        } catch (NotFoundException e) {
+            if (e.getObject().equals(User.class.getSimpleName())) {
+                log.error("When purchasing an item, user is not found for email {}", email);
+            } else {
+                log.error("When purchasing an item, store item is not found for id {}", storeItemId);
+            }
+            throw e;
+        } catch (ItemNotEnabledException e) {
+            log.error("Store item with id {} is not enabled for purchase", storeItemId);
+            throw e;
+        }  catch (InsufficientGemsException e) {
+            log.error("Not enough gems for user with email {}", email);
+            throw e;
+        } catch (Exception e) {
+            log.error("Failed to purchase item", e);
+            throw new SomethingWentWrongException("Failed to purchase item");
+        }
+    }
+
+    @Transactional
+    public RUserItem decreaseUserItemQuantity(String email, UUID storeItemId) throws Exception {
+        try {
+            User user = accountRepository.findUserByEmail(email)
+                    .orElseThrow(() -> new NotFoundException(User.class.getSimpleName(), true));
+
+            StoreItem storeItem = storeItemRepository.findById(storeItemId)
+                    .orElseThrow(() -> new NotFoundException(StoreItem.class.getSimpleName(), true));
+
+            UserItem userItem = userItemRepository.findByUserAndStoreItem(user, storeItem)
+                    .orElseThrow(() -> new NotFoundException(UserItem.class.getSimpleName(), true));
+
+            if (userItem.getQuantity() <= 0) {
+                throw new InsufficientUserItemQuantityException();
+            }
+
+            // Consume one unit of the user item
+            userItem.setQuantity(userItem.getQuantity() - 1);
+
+            if (userItem.getQuantity() <= 0) {
+                // If quantity becomes 0 or negative, delete the user item
+                userItemRepository.delete(userItem);
+            } else {
+                // Otherwise, save the updated user item
+                userItemRepository.save(userItem);
+            }
+
+            log.info("User item consumed successfully for user with email {}", email);
+            return RUserItem.builder().userItem(userItem).build();
+        } catch (NotFoundException e) {
+            if (e.getObject().equals(User.class.getSimpleName())) {
+                log.error("When decreasing user item quantity, user is not found for email {}", email);
+            } else if (e.getObject().equals(StoreItem.class.getSimpleName())) {
+                log.error("When decreasing user item quantity, store item is not found for id {}", storeItemId);
+            } else {
```

Comment:
This log logic could be more readable in a switch-case format, where the key is `e.getObject()`. Can you try?

Reply:
I tried that at first but it didn't work since switch-case statements need constant expressions, not getSimpleName(). 

Instead, we could do something such as:
`final String USER_CLASS_NAME = User.class.getSimpleName();` and then switch-case statements, but I felt that was a bit redundant. What do you think?

Reply:
I agree, we can leave it like this then